packages/xo-server/src/xapi/mixins/vm.mjs

```diff
@@ -140,53 +140,100 @@ const methods = {
       hasBootableDisk = true
     }
 
-    // Modify existing (previous template) disks if necessary
-    existingVdis &&
-      (await Promise.all(
-        mapToArray(existingVdis, async ({ size, $SR: srId, ...properties }, userdevice) => {
-          const vbd = find(vm.$VBDs, { userdevice })
-          if (!vbd) {
-            return
-          }
-          let vdi = vbd.$VDI
-          await this._setObjectProperties(vdi, properties)
+    if (existingVdis !== undefined) {
+      if (vdis === undefined) {
+        vdis = []
+      }
 
-          // if another SR is set, move it there
-          if (srId) {
-            vdi = await this.moveVdi(vdi.$id, srId)
-          }
+      forEach(existingVdis, ({ $SR: sr, ...properties }, userdevice) => {
+        const vbd = find(vm.$VBDs, { userdevice })
+        if (!vbd) {
+          return
+        }
 
-          // if the disk is bigger
-          if (size != null && size > vdi.virtual_size) {
-            await this.resizeVdi(vdi.$id, size)
-          }
-        })
-      ))
+        vdis.push({ ...properties, userdevice, sr })
+      })
+    }
 
-    // Creates the user defined VDIs.
-    //
-    // TODO: set vm.suspend_SR
-    if (!isEmpty(vdis)) {
+    if (vdis !== undefined && vdis.length > 0) {
       const devices = await this.call('VM.get_allowed_VBD_devices', vm.$ref)
-      await Promise.all(
-        mapToArray(vdis, (vdiDescription, i) =>
+      const _vdisToCreate = []
+      const _vdisToUpdate = []
+      const _vdisToDestroy = []
+
+      vdis.forEach(vdi => {
+        if (vdi.userdevice === undefined) {
+          vdi.userdevice = devices.shift()
+          _vdisToCreate.push(vdi)
+          return
+        }
+
+        // If the userdevice match no vbd, create the VDI
+        const vbd = find(vm.$VBDs, { userdevice: vdi.userdevice })
+        if (vbd === undefined) {
+          if (!vdi.destroy) {
+            const userdeviceIndex = devices.indexOf(vdi.userdevice)
+            if (userdeviceIndex === -1) {
```

Comment:
also, this is not really enforced by the xapi, you can really  create more disks than permitted by get_allowed_device